### PR TITLE
fix: prioritize full rollout over inplace restarts

### LINKED ISSUE #5401: [Bug]: replica cluster pod may be restarted twice on promotion
### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Contact Details

_No response_

### Version

trunk (main)

### What version of Kubernetes are you using?

1.31

### What is your Kubernetes environment?

Self-managed: kind (evaluation)

### How did you install the operator?

YAML manifest

### What happened?

Running the E2E (https://github.com/cloudnative-pg/cloudnative-pg/actions/runs/10612276341/job/29419506948) one of them failed. When the replica cluster is promoted, we may have two restarts on the replica cluster newly promoted single pod.

The first one is an in-place restart due to of a change in the configuration (archive mode), the second one is because the pod has to mount a new secret so it gets created again. This gets caught in the e2e and makes it fail, since after the cluster becomes ready it does additional checks and they fail because they find postgresql being restarted.

Discussing this with @leonardoce , we probably don't need to restart in place a pod that will have to be recreated anyway, since the recreation will also take care of the restart.

### Cluster resource

_No response_

### Relevant log output

```shell
2024-08-29T11:41:12.288663128Z stderr F {"level":"info","ts":"2024-08-29T11:41:12Z","msg":"Restarting primary instance in-place","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"ba7816c7-caf5-424e-bdc0-454c14d2a989","primaryPod":"cluster-replica-basicauth-1","reason":"Postgres needs a restart to apply some configuration changes"}
2024-08-29T11:41:12.300596376Z stderr F {"level":"debug","ts":"2024-08-29T11:41:12Z","msg":"Reconciliation loop end","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"ba7816c7-caf5-424e-bdc0-454c14d2a989","caller":"internal/controller/cluster_controller.go:134"}
2024-08-29T11:41:12.591002442Z stderr F {"level":"debug","ts":"2024-08-29T11:41:12Z","msg":"Reconciliation loop start","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"c9adb535-6145-4c3d-9dbf-0e5a02833483","caller":"internal/controller/cluster_controller.go:132"}
2024-08-29T11:41:12.60504582Z stderr F {"level":"debug","ts":"2024-08-29T11:41:12Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"c9adb535-6145-4c3d-9dbf-0e5a02833483","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:12.612076641Z stderr F {"level":"debug","ts":"2024-08-29T11:41:12Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"c9adb535-6145-4c3d-9dbf-0e5a02833483","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:12.669257776Z stderr F {"level":"info","ts":"2024-08-29T11:41:12Z","msg":"The current primary instance is fenced or is still recovering from it, we won't trigger a switchover","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"c9adb535-6145-4c3d-9dbf-0e5a02833483"}

[...]
2024-08-29T11:41:17.832661451Z stderr F {"level":"info","ts":"2024-08-29T11:41:17Z","msg":"The current primary instance is fenced or is still recovering from it, we won't trigger a switchover","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"9f509304-2f1d-4005-b934-8c026393e46d"}
2024-08-29T11:41:17.832839783Z stderr F {"level":"debug","ts":"2024-08-29T11:41:17Z","msg":"haven't found any instance to create","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"9f509304-2f1d-4005-b934-8c026393e46d","caller":"internal/controller/cluster_create.go:1311","instances":["cluster-replica-basicauth-1"],"dangling":[],"unusable":[]}
2024-08-29T11:41:17.832981516Z stderr F {"level":"debug","ts":"2024-08-29T11:41:17Z","msg":"Reconciliation loop end","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"9f509304-2f1d-4005-b934-8c026393e46d","caller":"internal/controller/cluster_controller.go:134"}
2024-08-29T11:41:18.700985865Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","msg":"Reconciliation loop start","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/controller/cluster_controller.go:132"}
2024-08-29T11:41:18.714894333Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:18.717999802Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:18.892801106Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","msg":"haven't found any instance to create","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/controller/cluster_create.go:1311","instances":["cluster-replica-basicauth-1"],"dangling":[],"unusable":[]}
2024-08-29T11:41:18.892818969Z stderr F {"level":"info","ts":"2024-08-29T11:41:18Z","msg":"Pod rollout required","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","podName":"cluster-replica-basicauth-1","reason":"original and target PodSpec differ in volumes: element app-secret has been added"}
2024-08-29T11:41:18.892823668Z stderr F {"level":"info","ts":"2024-08-29T11:41:18Z","msg":"Restarting primary instance without a switchover first","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","primaryPod":"cluster-replica-basicauth-1","reason":"original and target PodSpec differ in volumes: element app-secret has been added"}
2024-08-29T11:41:18.907993338Z stderr F {"level":"info","ts":"2024-08-29T11:41:18Z","msg":"Recreating instance pod","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","pod":"cluster-replica-basicauth-1","to":"ghcr.io/cloudnative-pg/postgresql:14.13","reason":"original and target PodSpec differ in volumes: element app-secret has been added"}
2024-08-29T11:41:18.908012394Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:18.908017684Z stderr F {"level":"info","ts":"2024-08-29T11:41:18Z","logger":"events","msg":"Upgrading instance cluster-replica-basicauth-1","type":"Normal","object":{"kind":"Cluster","namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","uid":"20917f53-d2d8-4915-896b-916a62eb4cfc","apiVersion":"postgresql.cnpg.io/v1","resourceVersion":"5091"},"reason":"UpgradingInstance"}
2024-08-29T11:41:18.918228727Z stderr F {"level":"debug","ts":"2024-08-29T11:41:18Z","msg":"Reconciliation loop end","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"df0037c5-0bde-446f-8e65-495c395c3968","caller":"internal/controller/cluster_controller.go:134"}
2024-08-29T11:41:19.882910346Z stderr F {"level":"debug","ts":"2024-08-29T11:41:19Z","msg":"Reconciliation loop start","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17","caller":"internal/controller/cluster_controller.go:132"}
2024-08-29T11:41:19.90221789Z stderr F {"level":"debug","ts":"2024-08-29T11:41:19Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:19.915281265Z stderr F {"level":"debug","ts":"2024-08-29T11:41:19Z","logger":"invokePlugin","msg":"correctly loaded the plugin client","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17","caller":"internal/cnpi/plugin/operatorclient/client.go:62"}
2024-08-29T11:41:20.034739425Z stderr F {"level":"info","ts":"2024-08-29T11:41:20Z","msg":"Waiting for the primary to be restarted without triggering a switchover","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17"}
2024-08-29T11:41:20.034888292Z stderr F {"level":"debug","ts":"2024-08-29T11:41:20Z","msg":"Instance pod not active. Retrying in one second.","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17","caller":"internal/controller/cluster_controller.go:641"}
2024-08-29T11:41:20.034970284Z stderr F {"level":"debug","ts":"2024-08-29T11:41:20Z","msg":"Cluster is in an in-place restart phase. Waiting...","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-replica-basicauth","namespace":"replica-mode-basic-auth-1346"},"namespace":"replica-mode-basic-auth-1346","name":"cluster-replica-basicauth","reconcileID":"6b2e9864-8fff-47a6-84f8-0f7721b0eb17","caller":"internal/controller/cluster_controller.go:647","phase":"Primary instance is being restarted without a switchover"}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mnencia**
Closes #5401

### CODE REVIEW

**Comment by gabriele-wolfox** [comment]
I would add a unit test to catch this change inside `internal/controller/cluster_upgrade_test.go` at line 107.

```suggestion

It("should prioritize full rollout over inplace restarts", func(ctx SpecContext) {
		pod := specs.PodWithExistingStorage(cluster, 1)

		status := postgres.PostgresqlStatus{
			Pod:            pod,
			IsPodReady:     true,
			ExecutableHash: "test_hash",
		}

		rollout := isInstanceNeedingRollout(ctx, status, &cluster)
		Expect(rollout.required).To(BeFalse())
		Expect(rollout.reason).To(BeEmpty())

		// Set a different image to trigger a full rollout
		pod.Spec.Containers[0].Image = "postgres:13.10"

		// Set pending restart to true in the status
		status = postgres.PostgresqlStatus{
			Pod:            pod,
			IsPodReady:     true,
			PendingRestart: true,
			ExecutableHash: "test_hash",
		}

		rollout = isInstanceNeedingRollout(ctx, status, &cluster)
		Expect(rollout.required).To(BeTrue())
		Expect(rollout.reason).ToNot(Equal("Postgres needs a restart to apply some configuration changes"))
	})
```

**Review by gabriele-wolfox — APPROVED** [review]

**Review by armru — APPROVED** [review]

File: internal/controller/cluster_upgrade_test.go [hunk 1]
```diff
@@ -105,6 +105,36 @@ var _ = Describe("Pod upgrade", Ordered, func() {
...
+			Pod:            pod,
+			IsPodReady:     true,
+			PendingRestart: true,
+			ExecutableHash: "test_hash",
+		}
+
+		rollout = isInstanceNeedingRollout(ctx, status, &cluster)
+		Expect(rollout.required).To(BeTrue())
+		Expect(rollout.canBeInPlace).To(BeFalse())
+		Expect(rollout.reason).To(Equal("the instance is using a different image: postgres:13.10 -> postgres:13.11"))
```
Comment (armru) [inline internal/controller/cluster_upgrade_test.go:135]: Given the deterministic nature of the test, I switched to an explicit error statement, any error reported back that is different is a cause of concern.

**Comment by fcanovai** [comment]
I've run a set of tests and the [run](https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/10683384461/job/29611574091) had one error. It seems to me that is an issue in the e2e where we do not wait for the cluster to be ready, but we immediately expect the ReadyInstances in the status to be equivalent to the number of instances, so we can end up failing if the cluster status is checked before it is updated. I'll try adding an additional check in the test.

**Comment by fcanovai** [comment]
I couldn't replicate a failure after the test fix neither in  https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/10685410791/job/29618298544 nor in https://github.com/cloudnative-pg/cloudnative-pg/actions/runs/10612276341/job/29419506948.

**Review by fcanovai — APPROVED** [review]

**Pull request merged**